### PR TITLE
feat(helm/provisioner): add support for provisioner keys, add note re psk

### PR DESCRIPTION
**Pull request by johnstcn**
Relates to https://github.com/coder/coder/issues/14985
- Adds `provisionerDaemon.keySecretName` and `provisionerDaemon.keySecretKey`
- Omitting `provisionerDaemon.pskSecretName` will now cause the PSK secret to no longer be created.
- Adds a note in `NOTES.txt` regarding provisioner PSKs.
- Adds validation that either `provisionerDaemon.keySecretName` or `provisionerDaemon.pskSecretName` is specified, and will fail the install in this case.

Manual smoke-testing:

1) With defaults:

```
helm install --namespace=tmp test-provisionerd . --set coder.image.tag=latest --dry-run 

NAME: test-provisionerd
LAST DEPLOYED: Thu Oct 17 14:45:55 2024
NAMESPACE: tmp
STATUS: pending-install
REVISION: 1
TEST SUITE: None
HOOKS:
MANIFEST: <yaml>

NOTES:
* Provisioner Daemon PSKs are no longer recommended for use with external
  provisioners. Consider migrating to scoped provisioner keys instead. For more
  information, see: https://coder.com/docs/admin/provisioners#authentication

Enjoy Coder! Please create an issue at https://github.com/coder/coder if you run
into any problems! :)

```

2) With explicitly specified `pskSecretName`:
```
helm install --namespace=tmp test-provisionerd . --set coder.image.tag=latest --set provisionerDaemon.pskSecretName='foobar' --dry-run

NAME: test-provisionerd
LAST DEPLOYED: Thu Oct 17 14:43:08 2024
NAMESPACE: tmp
STATUS: pending-install
REVISION: 1
TEST SUITE: None
HOOKS:
MANIFEST: <yaml>
NOTES:
* Provisioner Daemon PSKs are no longer recommended for use with external
  provisioners. Consider migrating to scoped provisioner keys instead. For more
  information, see: https://coder.com/docs/admin/provisioners#authentication

Enjoy Coder! Please create an issue at https://github.com/coder/coder if you run
into any problems! :)
```

3) With `pskSecretName` empty and `keySecretName` specified:
```
helm install --namespace=tmp test-provisionerd . --set coder.image.tag=latest --set provisionerDaemon.pskSecretName='' --set provisionerDaemon.keySecretName='foobar' --dry-run 

NAME: test-provisionerd
LAST DEPLOYED: Thu Oct 17 14:45:00 2024
NAMESPACE: tmp
STATUS: pending-install
REVISION: 1
TEST SUITE: None
HOOKS:
MANIFEST: <yaml>

NOTES:
Enjoy Coder! Please create an issue at https://github.com/coder/coder if you run
into any problems! :)

```

4) With `pskSecretName` empty and `keySecretName` empty: 

```
helm install --namespace=tmp test-provisionerd . --set coder.image.tag=latest --set provisionerDaemon.pskSecretName='' --set provisionerDaemon.keySecretName='' --dry-run

Error: INSTALLATION FAILED: execution error at (coder-provisioner/templates/coder.yaml:5:3): Either provisionerDaemon.pskSecretName or provisionerDaemon.keySecretName must be specified.


```

### CODE REVIEW

**Review by matifali — APPROVED** [review]

**Review by matifali — COMMENTED** [review]
LGTM.

File: helm/provisioner/tests/testdata/provisionerd_psk_and_key.golden [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        - start
+        command:
+        - /opt/coder
+        env:
+        - name: CODER_PROMETHEUS_ADDRESS
+          value: 0.0.0.0:2112
+        - name: CODER_PROVISIONER_DAEMON_PSK
+          valueFrom:
+            secretKeyRef:
+              key: psk
```
Comment (mtojek) [inline helm/provisioner/tests/testdata/provisionerd_psk_and_key.golden:113]: nit: any reason why this is called `psk` instead of `provisionerd-psk`?

**Review by mtojek — APPROVED** [review]

File: helm/provisioner/tests/testdata/provisionerd_psk_and_key.golden (hunk 1 above)
Comment (johnstcn) [reply helm/provisioner/tests/testdata/provisionerd_psk_and_key.golden:113]: The key name is currently hard-coded in `_coder.tpl`:

```
- name: CODER_PROVISIONER_DAEMON_PSK
  valueFrom:
    secretKeyRef:
      name: {{ .Values.provisionerDaemon.pskSecretName | quote }}
      key: psk
{{- if include "provisioner.tags" . }}
```

This wasn't chagned as part of this PR. I can add a separate PR to allow customizing the key name, if required.

File: helm/provisioner/tests/chart_test.go [hunk 2]
```diff
@@ -52,6 +52,18 @@ var testCases = []testCase{
 		name:          "provisionerd_psk",
 		expectedError: "",
 	},
+	{
+		name:          "provisionerd_key",
+		expectedError: "",
+	},
+	{
+		name:          "provisionerd_psk_and_key",
```
Comment (spikecurtis) [inline helm/provisioner/tests/chart_test.go:60]: Is it ever sensible to accept both? I think we can only accept one or other as the authentication credential

**Pull request merged**